GH-24868: [C++] Add a Tensor logical value type with varying dimensions, implemented using ExtensionType

rok opened the pull request:
### Rationale for this change

For use cases where underlying datatype and number of dimensions in tensors are equal but not the actual shape we want to add a `VariableShapeTensorType`.
See https://github.com/apache/arrow/issues/24868 and https://github.com/huggingface/datasets/issues/5272

### What changes are included in this PR?

This introduces definition of `arrow.variable_shape_tensor` extension and it's C++ implementation and a Python wrapper.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

This introduces new extension type to the user.
* Closes: #24868

westonpace commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,60 @@ Fixed shape tensor
+  plus optional dimension names with keys **"dim_names"** and ordering of
+  the dimensions with key **"permutation"**.
+
+  - Example: ``{ "ndim": 2}``
+  - Example with ``dim_names`` metadata for NCHW ordered data:
+
+    ``{ "ndim": 3, "dim_names": ["C", "H", "W"]}``
+
+  - Example of permuted 3-dimensional tensor:
+
+    ``{ "ndim": 3, "permutation": [2, 0, 1]}``
+
+    This is the physical layout shape and the shape of the logical
+    layout would given an individual tensor of shape [100, 200, 500]
+    be ``[500, 100, 200]``.
```
The way the description for this type (and for the fixed size tensor type actually) reads it makes it seem like there should be a `value_type` parameter in the JSON.  However, I think this parameter is actually expressed by the choice of storage type right?  E.g. if the storage type is `struct<data: List<Float64>, shape: FixedSizeList<3>>` then one can infer that the `value_type` is `Float64`.

However, couldn't you also infer that `ndim` is 3?

rok replied:
Good point! `ndim` is provided by storage type and serializing `ndim` would be redundant. I'll refactor it out of JSON.

westonpace commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,60 @@ Fixed shape tensor
+  * **data** is a ``List`` holding tensor elements of a single tensor.
+    Data type of the list elements is uniform across the entire column
+    and also provided in metadata.
+  * **shape** is a ``FixedSizeList`` of the tensor shape where
+    the size of the list is equal to the number of dimensions of the
+    tensor.
```
FixedSizeList of what?  Int32?  Int16?  Any integer type?

rok replied:
Indeed! Changed to:
```rst
  * **shape** is a ``FixedSizeList<uint32>[ndim]`` of the tensor shape where
    the size of the list ``ndim`` is equal to the number of dimensions of the
    tensor.
```

westonpace reviewed (commented):
I just looked through the `rst` portion and have a few questions.

rok commented:
Looking at [PyTorch `NestedTensor`](https://github.com/pytorch/pytorch/blob/3a3cf0e09d475df9237c95ebd14debf650e0c038/aten/src/ATen/native/nested/README.md#nestedtensors) it stores strides per tensor as opposed to our proposal which stores just a permutation for the entire array which can then be used to derive strides (using tensor shape) per tensor. In PyTorch if strides are not provided to the [constructor](https://github.com/pytorch/pytorch/blob/3c11184ca8fe35de70d1147f0f9d9beea8dc7e48/aten/src/ATen/NestedTensorImpl.cpp#L211C1-L221) they are apparently calculated assuming row-major/C-contiguous layout.
https://github.com/pytorch/pytorch/blob/3a3cf0e09d475df9237c95ebd14debf650e0c038/aten/src/ATen/NestedTensorImpl.cpp#L104C19-L130
This makes me assume strides are stored for caching (and not to allow variable data layout per tensor).

Should we adjust to store permutations (or strides) per row too or rather leave that for another extension (`GeneralVariableShapeTensorType`)? @lhoestq @mariosasko @AlenkaF @jorisvandenbossche @ezyang

lhoestq commented:
No strong opinion regarding storing strides. Cc @thomasw21 I'm not even sure that nested tensors are used that often no ? I'm also pinging other experts to get more insights.

AlenkaF commented:
How much extra complexity is added if we enable each tensor in an array to have it's own permutation defined?

rok commented:
@AlenkaF Storing permutations would allow per tensor memory layout, which I'm not sure is really needed in practice and could be confusing.
Storing strides (and perhaps requiring uniform permutation) could be useful as consumers wouldn't need to calculate strides when using tensors. Complexity wouldn't really increase.
Either way the size of type would increase by `n tensors x ndim x sizeof(int64)`.

ezyang commented:
This has been a contentious point for us internally. CC'ing @jbschlosser @cpuhrsch

If you want a generic container format that can represent anything that morally is a list of regular dense tensors, the most expressive way to do it is store individual strides per tensor. This is particularly useful if you have two jagged dimensions (e.g., list of varying size images on both height and width); if you store sizes/strides per batch this all just works fine.

However, it's a bit of a pain to store all the metadata, because most of the time you're not taking full advantage of all the expressivity; something a bit more compact will be more efficient, both in terms of memory usage and in terms of algorithmic efficiency. For example, it's common to have only one ragged dimension, so you have batch x ragged x some number of dense dimensions.

rok commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,76 @@ Fixed shape tensor
+  * **data** is a ``List`` holding tensor elements of a single tensor.
+    Data type of the list elements is uniform across the entire column
+    and also provided in metadata.
+  * **shape** is a ``FixedSizeList<uint32>[ndim]`` of the tensor shape where
+    the size of the list ``ndim`` is equal to the number of dimensions of the
+    tensor.
```
Are you saying you get `ndim` at query time and expect `ndim=const` over the entire array or are you saying that every row will potentially have different `ndim`? If latter is the case we could use storage:
```cpp
struct_({field("shape", list(uint32())), field("data", list(value_type))})
```
That would work if we can assume consistent memory layout (e.g. always row major). However if we want to allow for other memory layouts then we also need to store per row permutation or strides.

pitrou commented:
Hmm, is there a strong use case for `VariableDimensionTensor`?

rok commented:
@pitrou I imagine no, but perhaps something comes up.

pitrou commented:
Ah, well perhaps start with `VariableShapeTensor`? We don't need to do everything up front :-)

rok commented:
Seeing how ragged/jagged dimensions are used in [pytorch](https://pytorch.org/torchrec/torchrec.sparse.html#module-torchrec.sparse.jagged_tensor) and [tensorflow](https://www.tensorflow.org/api_docs/python/tf/RaggedTensor) I've added some language and an example [here](https://github.com/apache/arrow/pull/37166/files#diff-b54425cb176b53e51925c13a4d4e85cf7d03d4e1226e6d5bf4d7ae09923db8b3R197-R218). For now this would be purely metadata that would help converting arrow <-> jagged/ragged.

I'll also send the updated doc to the mailing list to see if there's more feedback to be had.

AlenkaF reviewed (commented):
The Python part LGTM. The only question that came to mind was if we want to add the `VariableShapeTensorScalar` class here also? Probably as a follow-up to not raise the complexity here.

mpelko commented:
As part of a group working a lot with point clouds (radar and lidar data), this is something we are really looking for. Specifically, we are using arrow for the storage of the training data and so far dumping point cloud data as binary blob (pyarrow.binary type).

All this to say `VariableShapeTensor` is *exactly* what we are looking for (in our case no need for `VariableDimensionTensor` so far).

Thanks for the effort and looking forward to this landing and making it into the release!

(happy to test or explain the use-case more if it helps in any way)

jorisvandenbossche commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,107 @@ Fixed shape tensor
+  tensor per row:
+
+  * **data** is a ``List`` holding tensor elements of a single tensor.
+    Data type of the list elements is uniform across the entire column
+    and also provided in metadata.
```
```suggestion
    Data type of the list elements is uniform across the entire column.
```

I don't think it is provided in the metadata? (and that is also not needed, since it is part of the type already)

jorisvandenbossche commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,107 @@ Fixed shape tensor
+* Description of the serialization:
+
+  The metadata must be a valid JSON object including number of
+  dimensions of the contained tensors as an integer with key **"ndim"**
```
Is this `ndim` needed? (you can infer it from the storage type, although it might be practical to include) 
If we keep it, I would also give a concrete example of minimal metadata as the first example in the list just below.

rok replied:
It's not needed and I'd omit it for now. I'll add a minimal example.

rok replied:
Changed.

rok commented:
Thanks for the review @jorisvandenbossche ! I've addressed your comments. Do you think the format proposal as is is ready to call a vote?

rok commented:
@pitrou do you think it'd be a good idea to split out the type definition into a separate PR or should we merge this in one piece?

pitrou commented:
> @pitrou do you think it'd be a good idea to split out the type definition into a separate PR or should we merge this in one piece?

Yes, it would be nice to have a separate PR for the Arrow spec additions.

pitrou commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,112 @@ Fixed shape tensor
+
+  * **data** is a ``List`` holding tensor elements of a single tensor.
+    Data type of the list elements is uniform across the entire column.
+  * **shape** is a ``FixedSizeList<uint32>[ndim]`` of the tensor shape where
```
I have not followed this, but was there any discussion of signed vs. unsigned here? Most of the Arrow spec uses signed integers even for numbers that are always positive (and unless your tensors are huge I'm not sure having one less bit to represent the shape's dimensions is a problem?).

rok replied:
I don't think this was discussed and indeed `int64` would be more consistent, will change.

jorisvandenbossche replied:
I think `int32` is also fine (just not `uint32`)? (we have other places where we used int32 by default for positive integers, eg the offsets)

pitrou commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,112 @@ Fixed shape tensor
+    in which case it is assumed that all dimensions might be variable.
+
+  * **uniform_shape** = shape of the dimensions that are guaranteed to stay
+    constant over all tensors in the array, with the shape of the ragged dimensions
```
The term "ragged" always feels obscure to me, how about "non-uniform"?

rok replied:
Since pytorch uses "jagged" and "tensorflow" ragged we might as well use "non-uniform" :D.
However the current text is being voted on, so if we change this we should probably communicate to the mailing list?

pitrou replied:
If they are just wording details then we don't need to inform the ML, IMHO.
However, if we make any semantic changes (such as the ``uniform_dimensions`` questions below), then it would need mentioning on the ML.

pitrou commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,112 @@ Fixed shape tensor
+    When logical and physical layout are equal, the permutation will always
+    be ([0, 1, .., N-1]) and can therefore be left out.
+
+  * **uniform_dimensions** = indices of dimensions whose sizes are
```
Here and below, clarify what happens if `uniform_dimensions` / `uniform_shape` is absent?

rok replied:
Added in `uniform_shape`.

pitrou commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,112 @@ Fixed shape tensor
+    ``{ "dim_names": ["C", "H", "W"] }``
+
+  - Example with ``uniform_dimensions`` metadata for a set of color images
+    with variable width:
```
Is it legal to pass `uniform_dimensions` but `uniform_shape`?

Come to think of it, perhaps we don't need two separate metadata members? For example, an absent `uniform_shape` would mean all dimensions are non-uniform. While a `null` entry in a present `uniform_shape` would point to a non-uniform dimension?

rok replied:
Indeed `uniform_dimensions` is a "weaker" version of `uniform_shape` information. We could have a case where we know dimensions of images won't change, but we don't know the shape until runtime. Could the information that dimension is uniform still be used for optimization at runtime? I imagine no. What do you think?

pitrou replied:
I have no idea. Perhaps potential users of this type do? :-)

pitrou commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,112 @@ Fixed shape tensor
+
+.. note::
+
+  With the exception of permutation all other parameters and storage
+  of VariableShapeTensor define the *physical* storage of the tensor.
```
```suggestion
  With the exception of ``permutation``, the parameters and storage
  of VariableShapeTensor relate to the *physical* storage of the tensor.
```

rok commented:
I've moved implementation to #38008 and changed the language to only ask for `uniform_shape` as discussed above.
@jorisvandenbossche & @pitrou could you please do a sanity check before we merge?

pitrou commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,109 @@ Fixed shape tensor
+
+    ``{}``
+
+  - Example with ``dim_names`` metadata for NCHW ordered data:
```
```suggestion
  - Example with ``dim_names`` metadata for NCHW ordered data (note that the first
    logical dimension, ``N``, is mapped to the **data** List array: each element in the List
    is a CHW tensor and the List of tensors implicitly constitutes a single NCHW tensor):
```

pitrou commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,109 @@ Fixed shape tensor
+
+    ``{ "dim_names": ["C", "H", "W"] }``
+
+  - Example with ``uniform_shape`` metadata for a set of color images
+    with variable width:
```
```suggestion
  - Example with ``uniform_shape`` metadata for a set of color images
    with fixed height, variable width and three color channels:
```

pitrou commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,109 @@ Fixed shape tensor
+
+    ``{ "permutation": [2, 0, 1] }``
+
+    This is the physical layout shape and the shape of the logical
+    layout would given an individual tensor of shape [100, 200, 500]
+    be ``[500, 100, 200]``.
```
```suggestion
    For example, if the physical **shape** of an individual tensor
    is ``[100, 200, 500]``, this permutation would denote a logical shape
    of ``[500, 100, 200]``.
```

pitrou commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,109 @@ Fixed shape tensor
+  With the exception of ``permutation``, the parameters and storage
+  of VariableShapeTensor relate to the *physical* storage of the tensor.
+
+  For example, consider a tensor with:
```
Nit: markup for a code block
```suggestion
  For example, consider a tensor with::
```

pitrou commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,109 @@ Fixed shape tensor
+  is composed of **data** and **shape** fields describing a single
+  tensor per row:
+
+  * **data** is a ``List`` holding tensor elements of a single tensor.
+    Data type of the list elements is uniform across the entire column.
```
```suggestion
  * **data** is a ``List`` holding tensor elements (each list element is
    a single tensor). The List's value type is the value type of the tensor,
    such as an integer or floating-point type.
```

pitrou commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,109 @@ Fixed shape tensor
+
+  This means the logical tensor has names [z, x, y] and shape [30, 10, 20].
+
+  Elements in a variable shape tensor extension array are stored
+  in row-major/C-contiguous order.
```
```suggestion
.. note::
   Values inside each **data** tensor element are stored in row-major/C-contiguous
   order according to the corresponding **shape**.
```

rok commented:
Thank for the suggestions @pitrou! If there's nothing further I'd tally the vote on ML and merge this.

pitrou commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,113 @@ Fixed shape tensor
+    non-uniform dimensions. This holds over all tensors in the array.
+    Sizes in uniform dimensions are represented with int32 values, while
+    sizes of the non-uniform dimensions are not known in advance and are
+    represented with 0s. If ``uniform_shape`` is not provided it is assumed
```
Or JSON supports `null` which maps quite naturally to what we are trying to express, IMHO.

rok replied:
Switched language to `null`.

pitrou commented:
@jorisvandenbossche Can I let you give this the final review?

rok commented:
@jorisvandenbossche can we merge this?

jorisvandenbossche commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,113 @@ Fixed shape tensor
+  key **"uniform_shape"**.
+  Minimal metadata is an empty JSON object.
+
+  - Example of minimal metadata is:
+
+    ``{}``
```
Sorry, one more small nitpick: the minimal metadata is actually _no_ metadata, which is typically represented as an empty string (I am actually not fully sure if in this case the metadata key could also just not be present in the field metadata), instead of an empty json dict (I don't think we should necessarily recommend using an empty dict)

rok replied:
That's a fair point! Empty string feels like the safer choice here. See my suggested change below.

rok commented on docs/source/format/CanonicalExtensions.rst:
```diff
@@ -148,6 +148,113 @@ Fixed shape tensor
+  with key **"permutation"**.
+  Shapes of tensors can be defined in a subset of dimensions by providing
+  key **"uniform_shape"**.
+  Minimal metadata is an empty JSON object.
+
+  - Example of minimal metadata is:
+
+    ``{}``
```
```suggestion
  Minimal metadata is an empty string.

  - In case of minimal metadata we assume that all dimensions are non-uniform
    and the ordering of dimensions is the same as the physical layout.
    Shape of individual tensors is given by ``shape`` ``FixedSizeList<int32>[ndim]`` 
    field which also implies number of dimensions with its length ``ndim``.
```

rok replied:
@jorisvandenbossche feel free to accept this suggestion or alter it.
22 @pitrou

rok replied:
Ping @jorisvandenbossche

jorisvandenbossche replied:
Personally I would say this paragraph is a bit superfluous, because that all follows from the main spec logically and from those parameters being optional?

rok replied:
I suppose so. So how about just:
```suggestion
  Minimal metadata is an empty string.
```

jorisvandenbossche commented:
Thanks a lot @rok !
